### PR TITLE
Update shared memory handling to work both in confined snaps and on pre-3.11 kernels

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Confined snaps only have access to "/dev/shm" via the "mir" interface and we want to deprecate that.

Wayland based toolkits typically use $XDG_RUNTIME_DIR to open shm pools, so we try that before "/dev/shm".

A confined snap that frigs with $XDG_RUNTIME_DIR (as we do for servers) will fail there and in "/dev/shm" so we try "/tmp" if both of the above fail.

Another way allocating a shm file can fail is on old kernels (as used on the current UBports phones), so while touching this code I've copied the workaround seen here:

https://github.com/MirServer/mir/blob/master/src/core/anonymous_shm_file.cpp#L54

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
I realize that verifying the "no O_TMPFILE" codepath is problematic. But this path is already broken, so I think we can merge if we're sure this doesn't make things worse.

File: examples/example-server-lib/wayland_helpers.cpp [hunk 1]
```diff
@@ -18,35 +18,72 @@
...
 #include <sys/mman.h>
 
 #include <cstring>
+#include <system_error>
 
 struct wl_shm_pool* make_shm_pool(struct wl_shm* shm, int size, void **data)
 {
-    struct wl_shm_pool *pool;
-    int fd;
+    static auto (*open_shm_file)() -> mir::Fd = []
```
[the remark this answers is not in the document]
Comment (RAOF) [reply examples/example-server-lib/wayland_helpers.cpp]: The compiler *should* be able to see that this is const anyway and not do something stupid like inserting the normal static mutex + initialiser thingy

[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply examples/example-server-lib/wayland_helpers.cpp]: You've misread the code. The lambda isn't called on initialization and the function pointer is reseated during the first call.

[the remark this answers is not in the document]
Comment (RAOF) [reply examples/example-server-lib/wayland_helpers.cpp]: Ah, I was wondering a little about that. In that case, maybe the code is too clever, Anna should just try the full set of options each time?

[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply examples/example-server-lib/wayland_helpers.cpp]: Maybe it's premature optimization, but hitting up to four filesystems to allocate shared memory seems excessive.

**Pull request merged**